tss-esapi/src/attributes/locality.rs

```diff
@@ -0,0 +1,127 @@
+
+    /// Returns the LocalityAttributes as a number.
+    ///
+    /// # Error
+    /// If the attributes are not extended en InvalidParams error
+    /// is returned.
+    pub fn as_extended(&self) -> Result<u8> {
+        if self.is_extended() {
+            Ok(self.0)
+        } else {
+            error!("Cannot retrieve LocalityAttributes as extended when the attributes are not indicated to be extended");
+            Err(Error::local_error(WrapperErrorKind::InvalidParam))
+        }
+    }
+
+    /// Returns the builder used to construct LocalAttributes.
+    pub fn builder() -> LocalityAttributesBuilder {
+        LocalityAttributesBuilder::new()
+    }
+}
+
+impl From<TPMA_LOCALITY> for LocalityAttributes {
+    fn from(tpma_locality: TPMA_LOCALITY) -> Self {
+        LocalityAttributes(tpma_locality)
+    }
+}
+
+impl From<LocalityAttributes> for TPMA_LOCALITY {
+    fn from(locality_attributes: LocalityAttributes) -> Self {
+        locality_attributes.0
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct LocalityAttributesBuilder {
+    localities: Vec<u8>,
+}
+
+impl LocalityAttributesBuilder {
+    /// Creates a new builder.
+    pub const fn new() -> Self {
+        LocalityAttributesBuilder {
+            localities: Vec::new(),
+        }
+    }
+    /// Adds a locality to the builder
+    pub fn with_locality(mut self, locality: u8) -> Self {
+        self.localities.push(locality);
+        self
+    }
+
+    /// Adds a slice of localities to the builder
+    pub fn with_localities(mut self, localities: &[u8]) -> Self {
+        self.localities.extend_from_slice(localities);
+        self
+    }
+
+    /// Builds the attributes
+    pub fn build(self) -> Result<LocalityAttributes> {
+        let mut locality_attributes = LocalityAttributes(0);
+        for locality in self.localities {
+            if locality_attributes.is_extended() {
+                error!("Failed to add locality {} to locality attributes because it contains an extended locality.", locality);
+                return Err(Error::local_error(WrapperErrorKind::InvalidParam));
+            }
+            match locality {
+                0 => locality_attributes.set_locality_zero(true),
+                1 => locality_attributes.set_locality_one(true),
+                2 => locality_attributes.set_locality_two(true),
+                3 => locality_attributes.set_locality_three(true),
+                4 => locality_attributes.set_locality_four(true),
+                5..=31 => {
+                    error!(
+                        "{} is an invalid locality and cannot be added to locality attributes",
+                        locality
+                    );
+                    return Err(Error::local_error(WrapperErrorKind::InvalidParam));
+                }
+                32.. => {
+                    if locality_attributes.0 != 0 {
+                        error!("Failed to add extended locality {} because locality attributes is not empty.", locality);
```

Comment:
Typo: "because locality attributes _are_ not empty"
:thinking: too bad that it's not possible to check if the correct log message has been produced in unit tests as there are several things that can return `WrapperErrorKind::InvalidParam` here.

Reply:
That would have been nice to check.

I am mixing things up. Sometimes the value can contain several attributes and sometimes it cannot. I have rewritten the messages hope they make better sense now.